runtime/drivers/athena/athena.go

```diff
@@ -0,0 +1,294 @@
+	return nil, false
+}
+
+func (c *Connection) AsFileStore() (drivers.FileStore, bool) {
+	return nil, false
+}
+
+// AsSQLStore implements drivers.Connection.
+func (c *Connection) AsSQLStore() (drivers.SQLStore, bool) {
+	return nil, false
+}
+
+// DownloadFiles returns a file iterator over objects stored in gcs.
+// The credential json is read from config google_application_credentials.
+// Additionally in case `allow_host_credentials` is true it looks for "Application Default Credentials" as well
+func (c *Connection) DownloadFiles(ctx context.Context, source *drivers.BucketSource) (drivers.FileIterator, error) {
+	conf, err := parseSourceProperties(source.Properties)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse config: %w", err)
+	}
+
+	prefix := "parquet_output_" + uuid.New().String()
+	bucketName := strings.TrimPrefix(strings.TrimRight(conf.OutputLocation, "/"), "s3://")
+	unloadPath := bucketName + "/" + prefix
+	err = c.unload(ctx, conf, "s3://"+unloadPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unload: %w", err)
+	}
+
+	bucketObj, err := c.openBucket(ctx, conf, bucketName)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open bucket %q: %w", unloadPath, err)
+	}
+
+	opts := rillblob.Options{
+		ExtractPolicy: &runtimev1.Source_ExtractPolicy{
+			// FilesStrategy: runtimev1.Source_ExtractPolicy_STRATEGY_HEAD,
+		},
+		GlobPattern: prefix + "/**",
+	}
+
+	it, err := rillblob.NewIterator(ctx, bucketObj, opts, c.logger)
+	if err != nil {
+		// TODO :: fix this for single file access. for single file first call only happens during download
+		var failureErr awserr.RequestFailure
+		if !errors.As(err, &failureErr) {
+			return nil, fmt.Errorf("failed to create the iterator %q %w", unloadPath, err)
+		}
+
+		// check again
+		if errors.As(err, &failureErr) && (failureErr.StatusCode() == http.StatusForbidden || failureErr.StatusCode() == http.StatusBadRequest) {
+			return nil, drivers.NewPermissionDeniedError(fmt.Sprintf("can't access remote err: %v", failureErr))
+		}
+	}
+
+	return it, err
+}
+
+func (c *Connection) openBucket(ctx context.Context, conf *sourceProperties, bucket string) (*blob.Bucket, error) {
+	cfg, err := awsconfig.LoadDefaultConfig(context.TODO(), awsconfig.WithSharedConfigProfile(conf.ProfileName))
+	if err != nil {
+		return nil, err
+	}
+
+	s3client := s3v2.NewFromConfig(cfg)
+	return s3blob.OpenBucketV2(ctx, s3client, bucket, nil)
+}
+
+func (c *Connection) unload(ctx context.Context, conf *sourceProperties, path string) error {
+	finalSQL := fmt.Sprintf("UNLOAD (%s) TO '%s' WITH (format = 'PARQUET')", conf.SQL, path)
+
+	cfg, err := awsconfig.LoadDefaultConfig(context.TODO(), awsconfig.WithSharedConfigProfile(conf.ProfileName))
+	if err != nil {
+		return err
+	}
+
+	client := athena.NewFromConfig(cfg)
+
+	resultConfig := &types.ResultConfiguration{
+		OutputLocation: aws.String("s3://" + strings.TrimPrefix(strings.TrimRight(conf.OutputLocation, "/"), "s3://") + "/output/"),
+	}
```

Comment:
Why is it necessary to specify the output location both here and in `UNLOAD`? Also, see note above about using URL parsing libraries if needing to manipulate URLs

Reply:
It's a required parameter for any query.
```
	// The location in Amazon S3 where your query and calculation results are stored,
	// such as s3://path/to/query/bucket/ . To run the query, you must specify the
	// query results location using one of the ways: either for individual queries
	// using either this setting (client-side), or in the workgroup, using
	// WorkGroupConfiguration . If none of them is set, Athena issues an error that no
	// output location is provided.
```